### PR TITLE
Remove Google Gson, replacing with Jackson Databind.  Upgrade Jackson Databind

### PR DESCRIPTION
**Pull request by tdonohue**
## Description
Removes Google Gson (JSON parser) from our dependency list, replacing it with Jackson Databind.  Most of our codebase is using Jackson Databind for JSON parsing (including our REST API), except for a handful of classes which use Gson as an alternative.   

This PR standardizes our codebase on Jackson Databind, allowing us to have one less dependency.

**NOTE: This PR also fixes CVE-2020-36518 by upgrading Jackson Databind to latest version.**

## Instructions for Reviewers
* Review the code changes. Most classes are either ITs or have ITs which validate the refactored code is correct.
* If possible @mspalti may want to re-test the IIIF `WordHighlightSolrSearch` class.  Unfortunately this class has _no existing tests_ and unclear documentation (I've been unable to get it to work on `main`, but it's likely cause of missing docs).  I **think** my updated code is accurate based on the old Gson code, but I haven't been able to test it.

### CODE REVIEW

**Comment by tdonohue** [comment]
@mspalti : This PR would be of interest to you as it refactors your IIIF `WordHighlightSolrSearch` class to no longer use Gson  (as that library is being removed).  Unfortunately, as noted in the description above, I don't know how to test this `WordHighlightSolrSearch` class as it has no existing tests & no documentation on how to get it working. I've been unsuccessful in getting it working even on `main`, so I'm hoping you would be able to more easily test that this PR doesn't break anything.

**Comment by mspalti** [comment]
@tdonohue , thanks, I'll test this asap. I should get it done by tomorrow (4/26) if not later today.

File: dspace-iiif/src/main/java/org/dspace/app/iiif/service/WordHighlightSolrSearch.java [hunk 1]
```diff
@@ -167,28 +160,36 @@ private SolrQuery getSolrQuery(String query, String manifestId) {
...
+        for (final JsonNode highEntry : highs) {
+            // Get the ocr_text node under the entry
+            JsonNode ocrNode = highEntry.get("ocr_text");
+            // Loop through the snippets array under that
+            for (final JsonNode snippet : ocrNode.get("snippets")) {
+                // Get a canvas ID based on snippet's pages
+                String pageId = getCanvasId(snippet.get("pages"));
+                // Loop through array of highlights for each snippet.
+                for (final JsonNode highlight : snippet.get("highlights")) {
+                    // Add annotation associated with each highlight
```
Comment (mspalti) [inline dspace-iiif/src/main/java/org/dspace/app/iiif/service/WordHighlightSolrSearch.java:191]: There can be multiple word highlights on a page so the the `highlight` JsonNode is an array. This addition worked or me:

```
for (int i = 0; i < highlight.size(); i++) {
        searchResult.addResource(getAnnotation(highlight.get(i), pageId, uuid));
}
```

**Review by mspalti — CHANGES_REQUESTED** [review]
@tdonohue if you add the array loop mentioned below then the changes to `WordHighlightSolrSearch` should work. I was able to query against the [Solr plugin](https://dbmdz.github.io/solr-ocrhighlighting/) and see results in Angular. Without the loop I was getting a 500.

**Comment by mspalti** [comment]
@tdonohue sorry this one is difficult to test! It's currently pretty "experimental." #8241 is an attempt to advance IIIF search support on several fronts and at the very least we should be able to add  some unit tests.

File: dspace-iiif/src/main/java/org/dspace/app/iiif/service/WordHighlightSolrSearch.java (hunk 1 above)
Comment (tdonohue) [reply dspace-iiif/src/main/java/org/dspace/app/iiif/service/WordHighlightSolrSearch.java]: Thanks @mspalti !  I've just pushed up this change into this PR.  If you could give it an approval (assuming it looks right), I'd appreciate it.

**Review by mspalti — APPROVED** [review]
I tested again with the changes. Looks good.

**Review by mwoodiupui — CHANGES_REQUESTED** [review]
Looks reasonable.  See inline comment about defending against nulls.  Otherwise OK by me.

File: dspace-iiif/src/main/java/org/dspace/app/iiif/service/WordHighlightSolrSearch.java [hunk 2]
```diff
@@ -167,27 +160,38 @@ private SolrQuery getSolrQuery(String query, String manifestId) {
...
-                    String pageId = getCanvasId(snippetArray.getAsJsonObject().get("pages"));
-                    for (JsonElement highlights : snippetArray.getAsJsonObject().getAsJsonArray("highlights")) {
-                        for (JsonElement highlight : highlights.getAsJsonArray()) {
-                            searchResult.addResource(getAnnotation(highlight, pageId, uuid));
-                        }
+
+        // Example structure of Solr response available at
+        // https://github.com/dbmdz/solr-ocrhighlighting/blob/main/docs/query.md
+        // Get the outer ocrHighlighting node
+        JsonNode highs = body.get("ocrHighlighting");
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-iiif/src/main/java/org/dspace/app/iiif/service/WordHighlightSolrSearch.java:179]: Good catch @mwoodiupui !  I added in null checks just now for all the newly added parsing code. If you could give this another look I'd appreciate it.

File: dspace-iiif/src/main/java/org/dspace/app/iiif/service/WordHighlightSolrSearch.java [hunk 3]
```diff
@@ -225,13 +234,16 @@ private AnnotationGenerator getAnnotation(JsonNode highlight, String pageId, UUI
...
+     * @return canvas id or null if node was null
      */
     private String getCanvasId(JsonNode pagesNode) {
-        JsonNode page = pagesNode.get(0);
-        String[] identArr = page.get("id").asText().split("\\.");
-        // the canvas id.
-        return "c" + identArr[1];
+        if (pagesNode != null) {
+            JsonNode page = pagesNode.get(0);
+            String[] identArr = page.get("id").asText().split("\\.");
```
Comment (mwoodiupui) [inline dspace-iiif/src/main/java/org/dspace/app/iiif/service/WordHighlightSolrSearch.java:242]: The previous line's `get` could return `null`, so `page` could be `null` at this point.

What a pain that `get` doesn't return an empty JsonNode.

**Review by mwoodiupui — CHANGES_REQUESTED** [review]
One last picky `null` issue.  Otherwise good.

File: dspace-iiif/src/main/java/org/dspace/app/iiif/service/WordHighlightSolrSearch.java (hunk 3 above)
Comment (mwoodiupui) [reply dspace-iiif/src/main/java/org/dspace/app/iiif/service/WordHighlightSolrSearch.java:242]: Come to think of it, line 242's `get` could also return `null`.

Comment (tdonohue) [reply dspace-iiif/src/main/java/org/dspace/app/iiif/service/WordHighlightSolrSearch.java]: @mwoodiupui : Ok, this is fixed.  There were two methods that used `get()` calls from JsonNode.  Patched them both in latest commit.

**Review by mwoodiupui — APPROVED** [review]

**Pull request merged**